src/vos/tests/vts_mvcc.c

```diff
@@ -0,0 +1,656 @@
+}
+
+static void
+set_akey(int i, char *path, daos_key_t *akey)
+{
+	int rc;
+
+	rc = snprintf(akey->iov_buf, akey->iov_buf_len, "%d-akey-%c", i,
+		      path[L_A]);
+	D_ASSERT(rc < akey->iov_buf_len);
+	akey->iov_len = strlen(akey->iov_buf) + 1;
+}
+
+static void
+set_value(int i, char *path, d_iov_t *value)
+{
+	int rc;
+
+	rc = snprintf(value->iov_buf, value->iov_buf_len, "%d-value-%s", i,
+		      path);
+	D_ASSERT(rc < value->iov_buf_len);
+	value->iov_len = strlen(value->iov_buf) + 1;
+}
+
+static int
+update_with_flags(struct io_test_args *arg, char *path, daos_epoch_t epoch,
+		  uint64_t flags)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	daos_unit_oid_t	 oid;
+	char		 dkey_buf[64];
+	daos_key_t	 dkey = {dkey_buf, sizeof(dkey_buf), 0};
+	char		 akey_buf[64];
+	daos_key_t	 akey = {akey_buf, sizeof(akey_buf), 0};
+	daos_iod_t	 iod;
+	char		 value_buf[64];
+	d_iov_t		 value = {value_buf, sizeof(value_buf), 0};
+	d_sg_list_t	 sgl;
+
+	set_oid(mvcc_arg->i, path, &oid);
+	set_dkey(mvcc_arg->i, path, &dkey);
+	set_akey(mvcc_arg->i, path, &akey);
+	set_value(mvcc_arg->i, path, &value);
+
+	memset(&iod, 0, sizeof(iod));
+	iod.iod_name = akey;
+	iod.iod_type = DAOS_IOD_SINGLE;
+	iod.iod_size = value.iov_len;
+	iod.iod_nr = 1;
+
+	memset(&sgl, 0, sizeof(sgl));
+	sgl.sg_nr = 1;
+	sgl.sg_iovs = &value;
+
+	flags |= VOS_OF_USE_TIMESTAMPS;
+
+	return vos_obj_update(arg->ctx.tc_co_hdl, oid, epoch, 0 /* pm_ver */,
+			      flags, &dkey, 1 /* iod_nr */, &iod, NULL, &sgl);
+}
+
+static int
+update_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return update_with_flags(arg, path, epoch, 0 /* flags */);
+}
+
+static int
+update_de_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return update_with_flags(arg, path, epoch, VOS_OF_COND_DKEY_INSERT);
+}
+
+static int
+update_dne_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return update_with_flags(arg, path, epoch, VOS_OF_COND_DKEY_UPDATE);
+}
+
+static int
+update_ae_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return update_with_flags(arg, path, epoch, VOS_OF_COND_AKEY_INSERT);
+}
+
+static int
+update_ane_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return update_with_flags(arg, path, epoch, VOS_OF_COND_AKEY_UPDATE);
+}
+
+static int
+puncho_with_flags(struct io_test_args *arg, char *path, daos_epoch_t epoch,
+		  uint64_t flags)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	daos_unit_oid_t	 oid;
+
+	set_oid(mvcc_arg->i, path, &oid);
+
+	flags |= VOS_OF_USE_TIMESTAMPS;
+
+	return vos_obj_punch(arg->ctx.tc_co_hdl, oid, epoch, 0 /* pm_ver */,
+			     flags, NULL /* dkey */, 0 /* akey_nr */,
+			     NULL /* akeys */, NULL /* dth */);
+}
+
+static int
+puncho_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
+{
+	return puncho_with_flags(arg, path, epoch, 0 /* flags */);
+}
+
+static int
+puncho_one_f(struct io_test_args *arg, char *path, daos_epoch_t epoch)
```

Comment:
See the comments in the table.

Reply:
"puncho" means "punch object", "one" means "if object is nonempty", "f" means "func" (as in "o_func" and "op_func_t"). These are shortened so that the table can be <= 80-char wide.